pkg/steps/release/promote.go

```diff
@@ -70,89 +64,129 @@ func (s *promotionStep) run(ctx context.Context) error {
-					return err
-				}
-				return fmt.Errorf("could not promote image streams: %w", err)
+func getImageMirror(config api.PromotionConfiguration, tags map[string]string, pipeline *imageapi.ImageStream) map[string]string {
+	if pipeline == nil {
+		return nil
+	}
+	imageMirror := map[string]string{}
+	if len(config.Name) > 0 {
+		for dst, src := range tags {
+			dockerImageReference := findDockerImageReference(pipeline, src)
+			if dockerImageReference == "" {
+				continue
+			}
+			dockerImageReference = getPublicImageReference(dockerImageReference, pipeline.Status.PublicDockerImageRepository)
+			imageMirror[dockerImageReference] = fmt.Sprintf("%s/%s/%s:%s", api.DomainForService(api.ServiceRegistry), config.Namespace, config.Name, dst)
+		}
+	} else {
+		for dst, src := range tags {
+			dockerImageReference := findDockerImageReference(pipeline, src)
+			if dockerImageReference == "" {
+				continue
 			}
-			return nil
-		})
+			dockerImageReference = getPublicImageReference(dockerImageReference, pipeline.Status.PublicDockerImageRepository)
+			imageMirror[dockerImageReference] = fmt.Sprintf("%s/%s/%s:%s", api.DomainForService(api.ServiceRegistry), config.Namespace, dst, config.Tag)
+		}
+	}
+	if len(imageMirror) == 0 {
+		return nil
 	}
+	return imageMirror
+}
 
-	client := s.dstClient.ImageStreamTags(s.config.Namespace)
-	for dst, src := range tags {
-		valid, _ := utils.FindStatusTag(pipeline, src)
-		if valid == nil {
-			continue
-		}
+func getPublicImageReference(dockerImageReference, publicDockerImageRepository string) string {
+	if !strings.Contains(dockerImageReference, ":5000") {
+		return dockerImageReference
+	}
+	i := strings.Index(publicDockerImageRepository, "/")
+	if i == -1 {
+		// This should never happen
+		log.Println("The status of the image stream Pipeline contains a publicDockerImageRepository which contains no / character.")
+		return dockerImageReference
+	}
+	publicHost := publicDockerImageRepository[:i]
+	i = strings.Index(dockerImageReference, "/")
+	if i == -1 {
+		// This should never happen
+		log.Println("The dockerImageReference contains no / character.")
+		return dockerImageReference
+	}
+	return fmt.Sprintf("%s%s", publicHost, dockerImageReference[i:])
+}
 
-		err := retry.RetryOnConflict(promotionRetry, func() error {
-			_, err := s.dstClient.ImageStreams(s.config.Namespace).Get(ctx, dst, meta.GetOptions{})
-			if errors.IsNotFound(err) {
-				_, err = s.dstClient.ImageStreams(s.config.Namespace).Create(ctx, &imageapi.ImageStream{
-					ObjectMeta: meta.ObjectMeta{
-						Name:      dst,
-						Namespace: s.config.Namespace,
-					},
-					Spec: imageapi.ImageStreamSpec{
-						LookupPolicy: imageapi.ImageLookupPolicy{
-							Local: true,
+func getPromotionPod(imageMirror map[string]string, namespace string) *coreapi.Pod {
+	var ocCommands []string
+	keys := make([]string, 0, len(imageMirror))
+	for k := range imageMirror {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
```

Comment:
We do this only so we get a deterministic order in tests?

Reply:
Exactly.